File: queuer/src/main/java/com/rndapp/task_feed/activities/LoginActivity.java
package com.rndapp.task_feed.activities;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.util.Log;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;

import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.google.gson.Gson;
import com.rndapp.task_feed.QueuerApplication;
import com.rndapp.task_feed.R;
import com.rndapp.task_feed.api.ServerCommunicator;
import com.rndapp.task_feed.models.ActivityUtils;
import com.rndapp.task_feed.models.User;
import org.json.JSONObject;

/**
 * Created with IntelliJ IDEA.
 * User: ell
 * Date: 7/13/13
 * Time: 2:12 PM
 */
public class LoginActivity extends Activity implements View.OnClickListener{
    private static final String USERNAME_KEY = "username";
    private static final String PASSWORED_KEY = "password";
    private static final String REMEMBER_KEY = "remember";

    private User user;

    EditText userField;
    EditText passField;
    CheckBox remember;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        SharedPreferences sp = getSharedPreferences(ServerCommunicator.API_KEY_PREFERENCE, Activity.MODE_PRIVATE);
        String apiKey = sp.getString("api_key", "");
        if (!apiKey.equals("")){
            startActivity(new Intent(this, FeedActivity.class));
            finish();
        }

        userField = (EditText)findViewById(R.id.username_field);
        passField = (EditText)findViewById(R.id.password_field);
        remember = (CheckBox)findViewById(R.id.remember_me_checkbox);

        String savedUsername = ActivityUtils.getUserCredential(this, USERNAME_KEY, "-1");
        String savedPassword = ActivityUtils.getUserCredential(this, PASSWORED_KEY, "-1");
        boolean saveCreds = ActivityUtils.getCredentialBoolean(this, REMEMBER_KEY, false);

        if (saveCreds){
            userField.setText(savedUsername);
            passField.setText(savedPassword);
        }

        remember.setChecked(saveCreds);

        findViewById(R.id.login_button).setOnClickListener(this);
        findViewById(R.id.create_account_button).setOnClickListener(this);
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()){
            case R.id.login_button:
                if (!userField.getText().toString().equals("") && !passField.getText().toString().equals("")){
                    findViewById(R.id.progress_bar).setVisibility(View.VISIBLE);
                    ServerCommunicator.login(this, ((QueuerApplication) getApplication()).getRequestQueue(),
                            userField.getText().toString(), passField.getText().toString(),
                            new Response.Listener<JSONObject>() {
                                boolean errored;
                                String errorText;

                                @Override
                                public void onResponse(JSONObject jsob) {
                                    try {
                                        Log.d("Received from /sessions", jsob.toString());
                                        if (jsob.has("api_key") && jsob.getString("api_key") != null) {
                                            String apiKey = jsob.getString("api_key");
                                            ActivityUtils.saveApiKey(LoginActivity.this, apiKey);
                                            //create user using Gson
                                            user = new Gson().fromJson(jsob.toString(), User.class);
                                            ActivityUtils.saveUserId(LoginActivity.this, user.getId());
                                        } else if (jsob.has("errors")) {
                                            //error
                                            errored = true;
                                            errorText = jsob.getString("errors");
                                        } else {
                                            //error
                                            errored = true;
                                            errorText = "Unknown Error";
                                        }
                                    } catch (Exception e) {
                                        e.printStackTrace();
                                        //error
                                        errored = true;
                                        errorText = e.getLocalizedMessage();
                                    }
                                    runOnUiThread(new Runnable() {
                                        @Override
                                        public void run() {
                                            findViewById(R.id.progress_bar).setVisibility(View.GONE);
                                            //go to the next activity
                                            if (!errored && user != null) {
                                                LoginActivity.this
                                                        .startActivity(new Intent(LoginActivity.this, FeedActivity.class));
                                                LoginActivity.this.finish();
                                            } else {
                                                ((TextView) findViewById(R.id.update)).setText(errorText);
                                            }
                                        }
                                    });
                                }
                            }, new Response.ErrorListener() {
                                @Override
                                public void onErrorResponse(final VolleyError volleyError) {
                                    runOnUiThread(new Runnable() {
                                        @Override
                                        public void run() {
                                            findViewById(R.id.progress_bar).setVisibility(View.GONE);
                                            ((TextView) findViewById(R.id.update)).setText(volleyError.getLocalizedMessage());
                                        }
                                    });

                                }
                            });

                    if (remember.isChecked()){
                        ActivityUtils.setCredentialBoolean(this, REMEMBER_KEY, true);
                        ActivityUtils.saveUserCredential(this, USERNAME_KEY, userField.getText().toString());
                        ActivityUtils.saveUserCredential(this, PASSWORED_KEY, passField.getText().toString());
                    }
                }
                break;
            case R.id.create_account_button:
                startActivity(new Intent(this, CreateAccountActivity.class));
                break;
        }
    }

}

File: queuer/src/main/java/com/rndapp/task_feed/activities/LoginActivity.kt
package com.rndapp.task_feed.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.TextUtils
import android.view.View
import android.widget.CheckBox
import android.widget.EditText
import android.widget.TextView
import com.android.volley.Response
import com.google.gson.Gson
import com.rndapp.task_feed.R
import com.rndapp.task_feed.api.LoginRequest
import com.rndapp.task_feed.api.VolleyManager
import com.rndapp.task_feed.managers.SessionManager
import com.rndapp.task_feed.models.SignInModel
import com.rndapp.task_feed.models.User
import org.json.JSONObject

/**
 * Created with IntelliJ IDEA.
 * User: ell
 * Date: 7/13/13
 * Time: 2:12 PM
 */
class LoginActivity : AppCompatActivity(), View.OnClickListener {

    private var user: User? = null

    internal var userField: EditText? = null
    internal var passField: EditText? = null
    internal var remember: CheckBox? = null

    internal var errored: Boolean = false
    internal var errorText: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        val sp = getSharedPreferences(SessionManager.API_KEY_PREFERENCE, Activity.MODE_PRIVATE)
        SessionManager.preferences = sp

        val apiKey = SessionManager.getApiKey()
        if (!TextUtils.isEmpty(apiKey)) {
            startActivity(Intent(this, SprintsActivity::class.java))
            finish()
        }

        userField = findViewById(R.id.username_field) as EditText
        passField = findViewById(R.id.password_field) as EditText
        remember = findViewById(R.id.remember_me_checkbox) as CheckBox

        val savedUsername = SessionManager.getUserCredential(this, USERNAME_KEY, "-1")
        val savedPassword = SessionManager.getUserCredential(this, PASSWORED_KEY, "-1")
        val saveCreds = SessionManager.getCredentialBoolean(this, REMEMBER_KEY, false)

        if (saveCreds) {
            userField!!.setText(savedUsername)
            passField!!.setText(savedPassword)
        }

        remember!!.isChecked = saveCreds

        findViewById(R.id.login_button).setOnClickListener(this)
        findViewById(R.id.create_account_button).setOnClickListener(this)
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.login_button -> if (userField!!.text.toString() != "" && passField!!.text.toString() != "") {
                findViewById(R.id.progress_bar).visibility = View.VISIBLE
                val signInModel = SignInModel()
                signInModel.username = userField!!.text.toString()
                signInModel.password = passField!!.text.toString()
                val request = LoginRequest(signInModel,
                        object : Response.Listener<JSONObject> {

                            override fun onResponse(jsob: JSONObject) {
                                handleApiKey(jsob)
                                handleLogin(jsob)
                            }
                        }, Response.ErrorListener { volleyError ->
                    runOnUiThread {
                        findViewById(R.id.progress_bar).visibility = View.GONE
                        (findViewById(R.id.update) as TextView).text = volleyError.localizedMessage
                    }
                })
                VolleyManager.queue?.add(request)

                if (remember!!.isChecked) {
                    SessionManager.setCredentialBoolean(this, REMEMBER_KEY, true)
                    SessionManager.saveUserCredential(this, USERNAME_KEY, userField!!.text.toString())
                    SessionManager.saveUserCredential(this, PASSWORED_KEY, passField!!.text.toString())
                }
            }
            R.id.create_account_button -> startActivity(Intent(this, CreateAccountActivity::class.java))
        }
    }

    fun handleApiKey(jsob: JSONObject) {
        try {
            if (jsob.has("api_key") && jsob.getString("api_key") != null) {
                val apiKey = jsob.getString("api_key")
                SessionManager.saveApiKey(this@LoginActivity, apiKey)
                //create user using Gson
                user = Gson().fromJson<User>(jsob.toString(), User::class.java!!)
                SessionManager.saveUserId(this@LoginActivity, user!!.id)
            } else if (jsob.has("errors")) {
                //error
                errored = true
                errorText = jsob.getString("errors")
            } else {
                //error
                errored = true
                errorText = "Unknown Error"
            }
        } catch (e: Exception) {
            e.printStackTrace()
            //error
            errored = true
            errorText = e.localizedMessage
        }
    }

    fun handleLogin(jsob: JSONObject) {
        runOnUiThread {
            findViewById(R.id.progress_bar).visibility = View.GONE
            //go to the next activity
            if (!errored && user != null) {
                this@LoginActivity
                        .startActivity(Intent(this@LoginActivity, SprintsActivity::class.java))
                this@LoginActivity.finish()
            } else {
                (findViewById(R.id.update) as TextView).text = errorText
            }
        }
    }

    companion object {
        private val USERNAME_KEY = "username"
        private val PASSWORED_KEY = "password"
        private val REMEMBER_KEY = "remember"
    }

}
